xdsl/irdl.py

```diff
@@ -727,6 +721,196 @@ class OptSuccessorDef(SuccessorDef, OptionalDef):
 OptSuccessor: TypeAlias = Block | None
 VarSuccessor: TypeAlias = list[Block]
 
+_ClsT = TypeVar("_ClsT")
+_ParamT = TypeVar("_ParamT")
+
+
+class OpDefField(Generic[_ClsT]):
+    cls: type[_ClsT]
+
+    def __init__(self, cls: type[_ClsT]):
+        self.cls = cls
+
+
+class ConstrainedOpDefField(Generic[_ClsT, _ParamT], OpDefField[_ClsT]):
+    param: _ParamT
+
+    def __init__(self, cls: type[_ClsT], param: _ParamT):
+        super().__init__(cls)
+        self.param = param
+
+
+class OperandFieldDef(
+    ConstrainedOpDefField[
+        OperandDef, AttrConstraint | Attribute | type[Attribute] | TypeVar
+    ]
+):
+    pass
+
+
+class ResultFieldDef(
+    ConstrainedOpDefField[
+        ResultDef, AttrConstraint | Attribute | type[Attribute] | TypeVar
+    ]
+):
+    pass
+
+
+class AttributeFieldDef(
+    ConstrainedOpDefField[
+        AttributeDef, AttrConstraint | Attribute | type[Attribute] | TypeVar
+    ]
+):
+    pass
+
+
+class RegionFieldDef(OpDefField[RegionDef]):
+    pass
+
+
+class SuccessorFieldDef(OpDefField[SuccessorDef]):
+    pass
+
+
+def result_def(
+    constraint: AttrConstraint | Attribute | type[Attribute] | TypeVar = Attribute,
+    *,
+    default: None = None,
```

Comment:
`dataclass` detects methods annotated with these three parameters, and uses them to inform things about the field. Any function with these argument labels can be used as a field definition.